eng/SourceBuildPrebuiltBaseline.xml

```diff
@@ -25,5 +25,8 @@
 
     <!-- Tracked in https://github.com/dotnet/source-build/issues/3438 -->
     <UsagePattern IdentityGlob="Microsoft.VisualStudio.Setup.Configuration.Interop/3.2.2146" />
+
+    <!-- MIBC -->
+    <UsagePattern IdentityGlob="optimization.*-*.mibc.runtime/*" />
```

Comment:
This seems a bit odd to me. Can you test these changes in the context of the VMR? I would expect that this will break source-build.

Reply:
We have a source build leg, which should probably show if something goes wrong? Also we can turn it off for SB if needed for the time being. We should look on what runtime is doing for the SB.

Reply:
The source-build leg in the PR runs for the local repo. This leg would probably fail with a prebuilt if you were to remove the line above. I'm suggesting copying over these changes to a local copy of the VMR and building there. I would expect to see a failure in that VMR build.

Essentially what I'm saying is that adding this line will likely have negative repercussions in the larger VMR build, and I'd like to avoid those.

Reply:
I'm assuming that it was already detected as a prebuilt which is what prompted adding it here.  It should not be added as an allowed prebuilt because when it flows to the SDK it will break source-build.  Ella's suggestion would prove that.  To be source-build compatible, these packages would need to be built as part of the source build.  If they cannot be entirely from source with an open source compatible license then they would be excluded.  Ideally the source-build product is as close as possible to the Microsoft product.  Optimization data is one area where they may differ.

Reply:
@ellahathaway @MichaelSimons So indeed, we added this line because SB was reporting the error in our CI.

Meanwhile, `dotnet/runtime` doesn't add these to exceptions, most likely because they work with them differently than F#. There are no other examples of using these packages in the dotnet org so no practice to get inspiration. My knowledge of VMR is limited, AFAIU there is no notion of any exclusion of these packages there.

What would be your advice? How hard would it be to build them as a part of source build? Not sure about the licence also, from what I know, the open source status of these packages is a bit unclear.

Reply:
Looking at the runtime usage of this package, it looks like it's being built by https://dev.azure.com/dnceng/internal/_git/dotnet-optimization. Based on this, I suspect that if we add entries for this package into the Version.Details.xml in fsharp, we can keep the exclusion in the baseline. I'll work to test this hypothesis and follow-up later today.

In the meantime, in case you encounter something like this in the future, I'll give you my long spiel on how I came to this conclusion and the general process SB takes for handling prebuilts:

> What would be your advice?

If source-build is detecting this as a prebuilt in the local repo build, the first course of action here is to determine if these changes also introduce a prebuilt into the VMR. There's a few ways to determine this: 1) clone dotnet/dotnet locally, make your changes to the fsharp repo in your local copy of dotnet/dotnet (at src/fsharp), then run

```bash
# Prep the source to build on your distro.
# This downloads a .NET SDK and a number of .NET packages needed to build .NET from source.
./prep-source-build.sh

# Build the .NET SDK
./build.sh -sb --clean-while-building
```

or 2) fork dotnet/dotnet, make your changes to fsharp repo on a branch (at src/fsharp), then open a draft PR against https://github.com/dotnet/dotnet. This will trigger a source-build PR check.

Frankly, I suspect that these changes in the current state will cause a prebuilt in the VMR. Under that assumption, there are a few things to consider:
1) Is this package currently being built by .NET? If yes, it's a candidate for source-build. In this case, add an entry for the package to Version.Details.xml, and re-build the VMR. If the prebuilt is no longer detected in the VMR build, great! This means that the package version is being lifted in the VMR and all is well. Add an exclusion for the package to SourceBuildPrebuiltBaseline.xml at the repo level.
2) Was this package previously being built by .NET? This package is likely a candidate for SBRP; SBRPs contain source buildable reference versions of historical .NET Core packages. If this package can be an SBRP, you'll have to open a PR in that repo to add the package then wait for the dependency flow to update in fsharp.
3) If this package comes from an open-source-repo outside of .NET, adding the repo to SBE may be an option.
4) If neither of the options above will work, we need to exclude this from source-build. Something as simple as conditioning the build and package reference will work.

> from what I know, the open source status of these packages is a bit unclear

From https://github.com/dotnet/arcade/blob/9b24668b7196b9639e60770465683cf81465d89a/Documentation/UnifiedBuild/VMR-Permissible-Sources.md#licenses: "The VMR does not permit code and binaries licensed under non-OSS licenses."

Reply:
@ellahathaway thanks a lot for your detailed guidance! I will follow the procedure at least to get my head around the topic. And thanks for looking into this concrete case as well, looking forward to your updates, this PR should be ready otherwise.

Reply:
Following up - these packages are non-OSS so they cannot be built from source. They will need to be excluded from source-build. I recommend conditioning on `DotNetBuildFromSource`.

Reply:
Thanks for all your assistance on that!
Alright then, I removed the exception from the baseline and conditioned it via DotNetBuildFromSource - this worked :)